### PR TITLE
tpm2_policypassword: Adding new tool to bind policies with PolicyPassword

### PR DESCRIPTION
**Pull request by idesai**

### CODE REVIEW

**Comment by idesai** [comment]
Has dependency on #1225

File: man/tpm2_policypassword.1.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+
+**tpm2_policypassword**(1) - Enables binding a policy to the authorization value of the authorized TPM object.
+
+# SYNOPSIS
+
+**tpm2_policypassword** [*OPTIONS*]
+
+# DESCRIPTION
+
+**tpm2_policypassword**(1)Enables binding a policy to the authorization value of the authorized TPM object. This is equivalent to authenticating using the object passphrase in plaintext, only this, enforces it as a policy. Requires extended session support using a resource manager. Also requires multi session support for satisfying the policy.
```
[the remark this answers is not in the document]
Comment (williamcroberts) [reply man/tpm2_policypassword.1.md]: Describe why you would want to use this. IIUC, by itself this doesn't give you much unless you want to start using policy boolean statements like and and or to say, "use a password or satisfy this other policy"

[the remark this answers is not in the document]
Comment (idesai) [reply man/tpm2_policypassword.1.md]: @williamcroberts, should then there be a requirements/pre-requisites section in the man page?

File: test/integration/tests/tcti/abrmd/policypassword.sh [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+tpm2_create -g sha256 -G aes -u $key_pub -r $key_priv -C $primary_key_ctx \
+  -L $policypassword -p $testpswd
+
+tpm2_load -C $primary_key_ctx -u $key_pub -r $key_priv -o $key_ctx
+tpm2_encryptdecrypt -c $key_ctx -o $encrypted_txt -I $plain_txt -p $testpswd
+
+tpm2_startauthsession -a -S $session_ctx
+tpm2_policypassword -S $session_ctx -o $policypassword
+tpm2_encryptdecrypt -c $key_ctx -I $encrypted_txt -o $decrypted_txt -D \
+  -p session:$session_ctx+$testpswd
```
Comment (williamcroberts) [inline test/integration/tests/tcti/abrmd/policypassword.sh]: Why do we need to do this `+$testpswd` if the tpm2_policypassword event has been generated? It needs to be in both the policyevent on session AND the authvalue in the command field?

Could you use the existing -p option to supply this rather then parsing it out?

File: man/tpm2_policypassword.1.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (williamcroberts) [reply man/tpm2_policypassword.1.md]: I think that's too far. I want to be able to comprehend what it does and why I want it when I read description. Explain it someone who is slightly TPM aware, but not savvy with it like you are. Ie dumb it down the best you can. Its a hard ask.

File: test/integration/tests/tcti/abrmd/policypassword.sh (hunk 2 above)
Comment (idesai) [reply test/integration/tests/tcti/abrmd/policypassword.sh]: Since we need to update both the authpolicy info and the authValue/hmac fields. So this would fail --
tpm2_encryptdecrypt -c key.ctx -I plain.txt -o enc.txt  -p session:session.ctx -p test
ERROR: Tss2_Sys_EncryptDecrypt2(0x12F) - tpm:error(2.0): authValue or authPolicy is not available for selected entity
ERROR: Unable to run tpm2_encryptdecrypt

File: man/tpm2_policypassword.1.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (idesai) [reply man/tpm2_policypassword.1.md]: I see. LMK if this is more readable now.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Imran Desai [mailto:notifications@github.com]
> Sent: Wednesday, November 7, 2018 9:13 AM
> To: tpm2-software/tpm2-tools <tpm2-tools@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-tools] tpm2_policypassword: Adding new tool
> to bind policies with PolicyPassword (#1227)
> 
> @idesai commented on this pull request.
> 
> ________________________________
> 
> In test/integration/tests/tcti/abrmd/policypassword.sh
> <https://github.com/tpm2-software/tpm2-
> tools/pull/1227#discussion_r231596406> :
> 
> > +tpm2_startauthsession -S $session_ctx
> +tpm2_policypassword -S $session_ctx -o $policypassword
> +tpm2_flushcontext -S $session_ctx
> +
> +tpm2_createprimary -a o -o $primary_key_ctx
> +
> +tpm2_create -g sha256 -G aes -u $key_pub -r $key_priv -C
> +$primary_key_ctx \
> +  -L $policypassword -p $testpswd
> +
> +tpm2_load -C $primary_key_ctx -u $key_pub -r $key_priv -o $key_ctx
> +tpm2_encryptdecrypt -c $key_ctx -o $encrypted_txt -I $plain_txt -p
> +$testpswd
> +
> +tpm2_startauthsession -a -S $session_ctx tpm2_policypassword -S
> +$session_ctx -o $policypassword tpm2_encryptdecrypt -c $key_ctx -I
> +$encrypted_txt -o $decrypted_txt -D \
> +  -p session:$session_ctx+$testpswd
> 
> Since we need to update both the authpolicy info and the authValue/hmac fields.
> So this would fail -- tpm2_encryptdecrypt -c key.ctx -I plain.txt -o enc.txt -p
> session:session.ctx -p test
> ERROR: Tss2_Sys_EncryptDecrypt2(0x12F) - tpm:error(2.0): authValue or
> authPolicy is not available for selected entity
> ERROR: Unable to run tpm2_encryptdecrypt

Don't the tools need to be smart enough to handle multiple -p calls so we can specify the 3 handles?

Then you just update the internal tool interface to take an array of auths?

File: man/tpm2_policypassword.1.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+# SYNOPSIS
+
+**tpm2_policypassword** [*OPTIONS*]
+
+# DESCRIPTION
+
+**tpm2_policypassword**(1) Enables a policy that requires the object's 
+authentication passphrase be provided. This is equivalent to authenticating 
+using the object passphrase in plaintext, only this, enforces it as a policy.
```
Comment (williamcroberts) [inline man/tpm2_policypassword.1.md]: no comma after `only this,`

File: man/tpm2_policypassword.1.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+**tpm2_policypassword** [*OPTIONS*]
+
+# DESCRIPTION
+
+**tpm2_policypassword**(1) Enables a policy that requires the object's 
+authentication passphrase be provided. This is equivalent to authenticating 
+using the object passphrase in plaintext, only this, enforces it as a policy. 
+It is a necessary way of requiring object's auth passphrase knowledge when the 
+object attributes mandate authorization only with policies; i.e. objects with 
+attribute "userwithauth = 0".
```
Comment (williamcroberts) [inline man/tpm2_policypassword.1.md]: This is awkward but I get it, try:

It provides a mechanism to allow for password authentication when an object only allows policy based authorization, ie object attribute "userwithauth" is 0.

File: man/tpm2_policypassword.1.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+Requires extended session support using a resource manager.
```
Comment (williamcroberts) [inline man/tpm2_policypassword.1.md]: This isn't 100% true, it also works if you **don't** use an RM, so the statement should be: If using a resource manager (RM), then one supporting extended sessions, like `[tpm2-abrmd](https://github.com/tpm2-software/tpm2-abrmd)` is required.

File: man/tpm2_policypassword.1.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+# OPTIONS
```
Comment (williamcroberts) [inline man/tpm2_policypassword.1.md]: new line above options

File: test/integration/tests/tcti/abrmd/policypassword.sh (hunk 2 above)
Comment (idesai) [reply test/integration/tests/tcti/abrmd/policypassword.sh]: @williamcroberts in this case though we are dealing with a single auth session.

**Comment by williamcroberts** [comment]
> -----Original Message-----
> From: Imran Desai [mailto:notifications@github.com]
> Sent: Wednesday, November 7, 2018 1:51 PM
> To: tpm2-software/tpm2-tools <tpm2-tools@noreply.github.com>
> Cc: Roberts, William C <william.c.roberts@intel.com>; Mention
> <mention@noreply.github.com>
> Subject: Re: [tpm2-software/tpm2-tools] tpm2_policypassword: Adding new tool
> to bind policies with PolicyPassword (#1227)
> 
> @idesai commented on this pull request.
> 
> ________________________________
> 
> In test/integration/tests/tcti/abrmd/policypassword.sh
> <https://github.com/tpm2-software/tpm2-
> tools/pull/1227#discussion_r231690863> :
> 
> > +tpm2_startauthsession -S $session_ctx
> +tpm2_policypassword -S $session_ctx -o $policypassword
> +tpm2_flushcontext -S $session_ctx
> +
> +tpm2_createprimary -a o -o $primary_key_ctx
> +
> +tpm2_create -g sha256 -G aes -u $key_pub -r $key_priv -C
> +$primary_key_ctx \
> +  -L $policypassword -p $testpswd
> +
> +tpm2_load -C $primary_key_ctx -u $key_pub -r $key_priv -o $key_ctx
> +tpm2_encryptdecrypt -c $key_ctx -o $encrypted_txt -I $plain_txt -p
> +$testpswd
> +
> +tpm2_startauthsession -a -S $session_ctx tpm2_policypassword -S
> +$session_ctx -o $policypassword tpm2_encryptdecrypt -c $key_ctx -I
> +$encrypted_txt -o $decrypted_txt -D \
> +  -p session:$session_ctx+$testpswd
> 
> @williamcroberts <https://github.com/williamcroberts>  in this case though we
> are dealing with a single auth session.

Ok so list of things could be 3 handles and an auth value... so the list contains 4 items,
Take the item that is an authvalue and shove it where it needs to go. The other
Three actual handles can go into the session array in order.

File: test/integration/tests/tcti/abrmd/policypassword.sh (hunk 2 above)
Comment (idesai) [reply test/integration/tests/tcti/abrmd/policypassword.sh]: as in, we are actually updating the auth value and session handle for same auth session.

Comment (williamcroberts) [reply test/integration/tests/tcti/abrmd/policypassword.sh]: @idesai I don't know why the command `tpm2_encryptdecrypt -c key.ctx -I plain.txt -o enc.txt -p session:session.ctx -p test` would work in the code base as it stands currently lookiung at your branch. It would need to be modified to track -p invocations in the option handler for the tool, build a list of up to 4 items, session handles and password, call the/an auth util helper, that converts the list of up to 4 strings to a list of up to 4 strings with at most 3 sessions and 1 authvalue.

File: man/tpm2_policypassword.1.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (williamcroberts) [reply man/tpm2_policypassword.1.md]: Way better

File: test/integration/tests/tcti/abrmd/policypassword.sh (hunk 2 above)
Comment (williamcroberts) [reply test/integration/tests/tcti/abrmd/policypassword.sh]: Hmm, for a TPM command their is only one field for an auth/hmac value.

**Comment by williamcroberts** [comment]
Just needs to pass travis

File: tools/tpm2_policypassword.c [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    }
+
+    TPM2B_DIGEST policy_digest = TPM2B_EMPTY_INIT;
+    result = tpm2_policy_get_digest(sapi_context, s, &policy_digest);
+    if (!result) {
+        LOG_ERR("Could not build tpm policy");
+        goto out;
+    }
+
+    tpm2_util_hexdump(policy_digest.buffer, policy_digest.size);
```
Comment (williamcroberts) [inline tools/tpm2_policypassword.c]: needs a newline output after hexdump.
